chain/postgresdb/database/database.go

```diff
@@ -0,0 +1,342 @@
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/jmoiron/sqlx"
+	"github.com/lib/pq" // Calls init function.
+
+	"github.com/drand/drand/log"
+)
+
+// lib/pq errorCodeNames
+// https://github.com/lib/pq/blob/master/error.go#L178
+const (
+	uniqueViolation = "23505"
+	undefinedTable  = "42P01"
+)
+
+// Set of error variables for CRUD operations.
+var (
+	ErrDBNotFound        = sql.ErrNoRows
+	ErrDBDuplicatedEntry = errors.New("duplicated entry")
+	ErrUndefinedTable    = errors.New("undefined table")
+)
+
+// Config is the required properties to use the database.
+type Config struct {
+	User         string
+	Password     string
+	Host         string
+	Name         string
+	MaxIdleConns int
+	MaxOpenConns int
+	DisableTLS   bool
+}
+
+// ConfigFromDSN provides support for creating a config from a DSN.
+func ConfigFromDSN(dsn string) (Config, error) {
+	conf, err := url.Parse(dsn)
+	if err != nil {
+		return Config{}, err
+	}
+	query := conf.Query()
+
+	password, _ := conf.User.Password()
+	cfg := Config{
+		User:       conf.User.Username(),
+		Password:   password,
+		Host:       conf.Host,
+		Name:       conf.Path,
+		DisableTLS: true,
+	}
+
+	if query.Has("sslmode") {
+		sslMode := query.Get("sslmode")
+		switch sslMode {
+		//nolint:goconst // Having to add constants is overkill.
+		case "disable":
+			cfg.DisableTLS = true
+		case "required":
+			cfg.DisableTLS = false
+		default:
+			return Config{}, fmt.Errorf("unsupported ssl mode %q. Expected disable or required", sslMode)
+		}
+	}
+
+	cfg.MaxIdleConns = 2
+	if query.Has("max-idle") {
+		max := query.Get("max-idle")
+		m, err := strconv.Atoi(max)
+		if err != nil {
+			return Config{}, fmt.Errorf("expected number for max-idle, got err: %w", err)
+		}
+		if m >= 0 {
+			cfg.MaxIdleConns = m
+		}
+	}
+
+	cfg.MaxOpenConns = 0
+	if query.Has("max-open") {
+		open := query.Get("max-open")
+		o, err := strconv.Atoi(open)
+		if err != nil {
+			return Config{}, fmt.Errorf("expected number for max-open, got err: %w", err)
+		}
+		if o >= 0 {
+			cfg.MaxOpenConns = o
+		}
+	}
+
+	return cfg, nil
+}
+
+// Open knows how to open a database connection based on the configuration.
+// It also performs a health check to make sure the connection is healthy.
+//
+//nolint:gocritic // There is nothing wrong with using value semantics here.
+func Open(ctx context.Context, cfg Config) (*sqlx.DB, error) {
+	sslMode := "require"
+	if cfg.DisableTLS {
+		sslMode = "disable"
+	}
+
+	q := make(url.Values)
+	q.Set("sslmode", sslMode)
+	q.Set("timezone", "utc")
+
+	u := url.URL{
+		Scheme:   "postgres",
+		User:     url.UserPassword(cfg.User, cfg.Password),
+		Host:     cfg.Host,
+		Path:     strings.ToLower(cfg.Name),
+		RawQuery: q.Encode(),
+	}
+
+	db, err := sqlx.Open("postgres", u.String())
+	if err != nil {
+		return nil, err
+	}
+	db.SetMaxIdleConns(cfg.MaxIdleConns)
+	db.SetMaxOpenConns(cfg.MaxOpenConns)
+
+	return db, StatusCheck(ctx, db)
```

Comment:
It's called with a `context.WithTimeout()`. However, it might be a good idea to have its own timeout here.
What do you think? Would 5 seconds be reasonable?